### DISCUSSION TITLE
What's the best way to prevent dependency confusion attack with a private source?

### DISCUSSION THREAD

**Discussion by ysk24ok** [python-poetry/poetry#8308] [Q&A] [answered]
Hello community.
We have a package `foo` in our private source and the package depends on another package `subfoo` in the same private source.
I want Poetry not to access PyPI to install those libraries to prevent dependency confusion attack.
To achieve this goal, I think there are two ways.

1. Use `explicit`
    ```
    [[tool.poetry.source]]
    name = "private"
    url = ...
    priority = "explicit"

    [tool.poetry.dependencies]
    foo = {source = "private", version = "*"}
    subfoo = {source = "private", version = "*"}
    ```
    - This looks OK, but it's a bit troublesome to put all private packages in the tool.poetry.dependencies section.
1. Set higher priority than PyPI
    ```
    [[tool.poetry.source]]
    name = "private"
    url = ...
    priority = "default"

    [[tool.poetry.source]]
    name = "PyPI"
    priority = "primary"

    [tool.poetry.dependencies]
    foo = {version = "*"}
    ```
    - This way IIUC Poetry accesses the private source first and if it can't find the package it falls back to PyPI.
    - I don't have to put `subfoo` package and also don't have to specify `source` for `foo` package.
    - But I'm wondering which priorities I should use? I think there are several combinations.
        - private: default, PyPI: primary
        - private: default, PyPI: supplemental
        - private: primary, PyPI: supplemental

I'd appreciate your help. Thank you in advance!

**Comment by radoering** [comment, marked as answer]
The solution with explicit sources has the advantage that only one source is searched for each dependency (better performance of `poetry lock`). I'd prefer that solution if most dependencies come from PyPI and there are not too many transitive private dependencies.

Re 1:

I'd put the transitive dependencies in a separate optional group. That way they are still considered during dependency resolution but not installed by default if they are not required anymore by any other dependency. Of course, you'll still want to remove them if they are not required anymore but IMO it feels a bit cleaner at least.

```
[tool.poetry.dependencies]
foo = {source = "private", version = "*"}

[tool.poetry.group.transitive]
optional = true

[tool.poetry.group.transitive.dependencies]
subfoo = {source = "private", version = "*"}
```

Re 2:

"private: default, PyPI: supplemental" and "private: primary, PyPI: supplemental" are equivalent. "Default" just disables the implicit PyPI source, which does not matter if you specify it explicitly, and is equivalent to "first primary". "private: default, PyPI: primary" will always query both sources even if packages were found in the first source. That's probably not what you want.

**Reply by ysk24ok** [reply]
@radoering Thank you for your detailed answer!
As I have a few packages in our private source, using `explicit` would be better.

> I'd put the transitive dependencies in a separate optional group.

Didn't know we can specify transitive dependencies but in our case `foo` requires `subfoo` to be always installed.

> "private: default, PyPI: primary" will always query both sources even if packages were found in the first source. That's probably not what you want.

Exactly, that's not what I want.

**Reply by ysk24ok** [reply]
>  in our case foo requires subfoo to be always installed.

Thinking twice, I noticed using `transitive` group would make it clearer that `subfoo` is not a direct dependency as you said.

```
[tool.poetry.dependencies]
foo = {source = "private", version = "*"}

[tool.poetry.group.transitive.dependencies]
subfoo = {source = "private", version = "*"}
```

**Reply by ysk24ok** [reply]
Follow-up: I tried the use of `transitive` group, but I encountered an error saying `Package subfoo (X.Y.Z) not found.` when e.g. `poetry update` is executed and Poetry tried to fetch `foo`. I suspect Poetry tried to fetch `subfoo` from PyPI as `main` group doesn't have `subfoo` package. I ended up deleting `transitive` group and add `subfoo` to `main`.
